File: src/agent_dump/collect_reduction.py
"""Concurrent session summarization and tree reduction for collect mode."""

from collections import defaultdict
from collections.abc import Callable
from concurrent.futures import ThreadPoolExecutor
from datetime import date, tzinfo
import sys
import threading

from agent_dump.bounded_concurrency import iter_completed_futures
from agent_dump.collect_logging import CollectLogger
from agent_dump.collect_models import (
    GROUP_SIZE,
    MAX_SUMMARY_ITEMS_PER_FIELD,
    SESSION_MERGE_LLM_THRESHOLD,
    CollectAggregate,
    CollectMode,
    CollectProgressEvent,
    CollectSummaryGroup,
    MergeSessionsProgress,
    PlannedCollectEntry,
    SessionSummaryEntry,
    StructuredSummaryContext,
    StructuredSummaryPhase,
    SummarizeChunksProgress,
    TreeReductionProgress,
)
from agent_dump.collect_progress import emit_collect_progress
from agent_dump.collect_prompts import build_collect_chunk_prompt, build_collect_merge_prompt
from agent_dump.collect_requests import StructuredSummaryRequester, request_structured_summary_from_llm
from agent_dump.collect_summary import (
    merge_summary_payloads,
    summary_payload_size,
)
from agent_dump.config import MAX_COLLECT_SUMMARY_CONCURRENCY, AIConfig
from agent_dump.i18n import Keys, i18n
from agent_dump.terminal_output import render_terminal_message


def _summary_needs_compression(payload: dict[str, list[str]]) -> bool:
    return summary_payload_size(payload) > SESSION_MERGE_LLM_THRESHOLD or any(
        len(items) > MAX_SUMMARY_ITEMS_PER_FIELD for items in payload.values()
    )


def _summarize_collect_entry(
    *,
    config: AIConfig,
    planned_entry: PlannedCollectEntry,
    timeout_seconds: int,
    local_tz: tzinfo | None,
    on_chunk_summarized: Callable[[SummarizeChunksProgress], None] | None = None,
    on_session_merged: Callable[[MergeSessionsProgress], None] | None = None,
    logger: CollectLogger | None = None,
    mode: CollectMode = CollectMode.PM,
    request_structured_summary: StructuredSummaryRequester = request_structured_summary_from_llm,
) -> SessionSummaryEntry:
    entry = planned_entry.collect_entry
    chunks = planned_entry.chunks
    chunk_payloads: list[dict[str, list[str]]] = []
    for chunk_index, chunk_events in enumerate(chunks):
        prompt = build_collect_chunk_prompt(
            entry,
            chunk_events,
            chunk_index=chunk_index,
            chunk_total=len(chunks),
            local_tz=local_tz,
            mode=mode,
        )
        payload = request_structured_summary(
            config,
            prompt,
            context=StructuredSummaryContext(
                label=f"{entry.session_uri} chunk {chunk_index + 1}/{len(chunks)}",
                phase=StructuredSummaryPhase.CHUNK_SUMMARY,
                session_uri=entry.session_uri,
                chunk_index=chunk_index + 1,
                chunk_total=len(chunks),
            ),
            timeout_seconds=timeout_seconds,
            logger=logger,
            mode=mode,
        )
        chunk_payloads.append(payload)
        emit_collect_progress(
            on_chunk_summarized,
            SummarizeChunksProgress(
                current=1,
                total=1,
                concurrency=1,
                session_uri=entry.session_uri,
                chunk_index=chunk_index + 1,
                chunk_total=len(chunks),
            ),
        )

    merged = merge_summary_payloads(chunk_payloads, max_items_per_field=None, mode=mode)
    if len(chunk_payloads) > 1 and _summary_needs_compression(merged):
        try:
            merged = request_structured_summary(
                config,
                build_collect_merge_prompt(
                    source_uri=entry.session_uri,
                    payloads=chunk_payloads,
                    merge_label="session",
                    mode=mode,
                ),
                context=StructuredSummaryContext(
                    label=f"{entry.session_uri} session merge",
                    phase=StructuredSummaryPhase.SESSION_MERGE,
                    session_uri=entry.session_uri,
                    chunk_total=len(chunks),
                ),
                timeout_seconds=timeout_seconds,
                logger=logger,
                mode=mode,
            )
        except RuntimeError as exc:
            if logger is not None:
                logger.log(
                    "llm_merge_fallback",
                    phase=StructuredSummaryPhase.SESSION_MERGE.value,
                    session_uri=entry.session_uri,
                    chunk_total=len(chunks),
                    error=str(exc),
                )
    emit_collect_progress(
        on_session_merged,
        MergeSessionsProgress(
            current=1,
            total=1,
            session_uri=entry.session_uri,
            chunk_total=len(chunks),
        ),
    )

    return SessionSummaryEntry(
        collect_entry=entry,
        summary_data=merged,
    )


def summarize_collect_entries(
    *,
    config: AIConfig,
    planned_entries: list[PlannedCollectEntry],
    summary_concurrency: int,
    local_tz: tzinfo | None = None,
    progress_callback: Callable[[CollectProgressEvent], None] | None = None,
    timeout_seconds: int = 90,
    logger: CollectLogger | None = None,
    mode: CollectMode = CollectMode.PM,
    request_structured_summary: StructuredSummaryRequester = request_structured_summary_from_llm,
) -> list[SessionSummaryEntry]:
    """Generate structured per-session summaries with limited concurrency."""
    if not planned_entries:
        return []

    total = len(planned_entries)
    total_chunks = sum(len(item.chunks) for item in planned_entries)
    max_workers = min(max(1, summary_concurrency), MAX_COLLECT_SUMMARY_CONCURRENCY)
    results: list[SessionSummaryEntry | None] = [None] * total
    chunk_progress_lock = threading.Lock()
    merge_progress_lock = threading.Lock()
    summarized_chunks = 0
    merged_sessions = 0
    failed_sessions = 0
    last_error: Exception | None = None

    emit_collect_progress(
        progress_callback,
        SummarizeChunksProgress(current=0, total=total_chunks, concurrency=max_workers),
    )
    emit_collect_progress(
        progress_callback,
        MergeSessionsProgress(current=0, total=total),
    )

    def _mark_chunk_summarized(event: SummarizeChunksProgress) -> None:
        nonlocal summarized_chunks
        with chunk_progress_lock:
            summarized_chunks += event.current
            current = summarized_chunks
        emit_collect_progress(
            progress_callback,
            SummarizeChunksProgress(
                current=current,
                total=total_chunks,
                session_uri=event.session_uri,
                chunk_index=event.chunk_index,
                chunk_total=event.chunk_total,
                concurrency=max_workers,
            ),
        )

    def _mark_session_merged(event: MergeSessionsProgress) -> None:
        nonlocal merged_sessions
        with merge_progress_lock:
            merged_sessions += event.current
            current = merged_sessions
        emit_collect_progress(
            progress_callback,
            MergeSessionsProgress(
                current=current,
                total=total,
                session_uri=event.session_uri,
                chunk_total=event.chunk_total,
            ),
        )

    def _summarize(planned_entry: PlannedCollectEntry) -> SessionSummaryEntry:
        return _summarize_collect_entry(
            config=config,
            planned_entry=planned_entry,
            timeout_seconds=timeout_seconds,
            local_tz=local_tz,
            on_chunk_summarized=_mark_chunk_summarized,
            on_session_merged=_mark_session_merged,
            logger=logger,
            mode=mode,
            request_structured_summary=request_structured_summary,
        )

    with ThreadPoolExecutor(max_workers=max_workers) as executor:
        for index, planned_entry, future in iter_completed_futures(
            planned_entries,
            max_pending=max_workers,
            submit=lambda item: executor.submit(_summarize, item),
        ):
            try:
                results[index] = future.result()
            except Exception as exc:  # noqa: BLE001
                failed_sessions += 1
                last_error = exc
                entry = planned_entry.collect_entry
                print(
                    render_terminal_message(
                        Keys.WARN_SESSION_SUMMARY_SKIPPED,
                        uri=entry.session_uri,
                        error=exc,
                    ),
                    file=sys.stderr,
                )
                if logger is not None:
                    logger.log(
                        "session_summary_failed",
                        session_uri=entry.session_uri,
                        error=str(exc),
                    )

    summaries = [item for item in results if item is not None]
    if not summaries and last_error is not None:
        raise last_error
    if failed_sessions:
        print(i18n.t(Keys.WARN_SESSION_SUMMARY_FAILURES, count=failed_sessions), file=sys.stderr)
    return summaries


def reduce_collect_summaries(
    *,
    config: AIConfig,
    session_summaries: list[SessionSummaryEntry],
    timeout_seconds: int = 90,
    group_size: int = GROUP_SIZE,
    progress_callback: Callable[[CollectProgressEvent], None] | None = None,
    logger: CollectLogger | None = None,
    mode: CollectMode = CollectMode.PM,
    request_structured_summary: StructuredSummaryRequester = request_structured_summary_from_llm,
) -> CollectAggregate:
    """Reduce summaries only within the attribution required by the report mode."""
    working = [
        CollectSummaryGroup(
            date_value=summary.collect_entry.date_value,
            project_directory=summary.collect_entry.project_directory,
            session_uris=(summary.collect_entry.session_uri,),
            summary_data=summary.summary_data,
        )
        for summary in session_summaries
    ]
    reduction_depth = 0

    while working:
        grouped: dict[tuple[date, str, str], list[CollectSummaryGroup]] = defaultdict(list)
        for group in working:
            scope = group.session_uris[0] if mode is CollectMode.INSIGHT or not group.project_directory else ""
            grouped[(group.date_value, group.project_directory, scope)].append(group)
        if len(grouped) == len(working):
            break

        reduction_depth += 1
        next_level: list[CollectSummaryGroup] = []
        total_groups = sum((len(groups) + group_size - 1) // group_size for groups in grouped.values())
        emit_collect_progress(
            progress_callback,
            TreeReductionProgress(level=reduction_depth, current=0, total=total_groups),
        )
        group_index = 0
        for groups in grouped.values():
            for start in range(0, len(groups), group_size):
                batch = groups[start : start + group_size]
                group_index += 1
                first = batch[0]
                payloads = [group.summary_data for group in batch]
                merged = merge_summary_payloads(payloads, max_items_per_field=None, mode=mode)
                group_source = f"collect://group-level-{reduction_depth}/group-{group_index}"
                if len(batch) > 1 and _summary_needs_compression(merged):
                    try:
                        merged = request_structured_summary(
                            config,
                            build_collect_merge_prompt(
                                source_uri=group_source,
                                payloads=payloads,
                                merge_label=f"group-level-{reduction_depth}",
                                mode=mode,
                            ),
                            context=StructuredSummaryContext(
                                label=group_source,
                                phase=StructuredSummaryPhase.GROUP_MERGE,
                            ),
                            timeout_seconds=timeout_seconds,
                            logger=logger,
                            mode=mode,
                        )
                    except RuntimeError as exc:
                        if logger is not None:
                            logger.log(
                                "llm_merge_fallback",
                                phase=StructuredSummaryPhase.GROUP_MERGE.value,
                                context=group_source,
                                level=reduction_depth,
                                group_index=group_index,
                                error=str(exc),
                            )
                next_level.append(
                    CollectSummaryGroup(
                        date_value=first.date_value,
                        project_directory=first.project_directory,
                        session_uris=tuple(uri for group in batch for uri in group.session_uris),
                        summary_data=merged,
                    )
                )
                emit_collect_progress(
                    progress_callback,
                    TreeReductionProgress(level=reduction_depth, current=group_index, total=total_groups),
                )
        working = next_level

    return CollectAggregate(groups=tuple(working), reduction_depth=reduction_depth)

File: src/collect_reduction.rs
use crate::collect_log::Logger;
use crate::collect_model::{Entry, Group, Mode, Summary};
use crate::collect_progress::progress;
use crate::collect_summary::{self, Context};
use crate::config::{AiConfig, CollectConfig};
use serde_json::json;
use std::collections::HashMap;
use std::io::Write;
use std::sync::{
    atomic::{AtomicUsize, Ordering},
    mpsc,
};

enum Event {
    Chunk,
    Complete(usize, crate::Result<Summary>),
}

fn summarize(
    entry: &Entry,
    ai: &AiConfig,
    timeout: u64,
    mode: Mode,
    logger: &Logger,
    sender: &mpsc::Sender<Event>,
) -> crate::Result<Summary> {
    let uri = entry.uri();
    let mut payloads = Vec::new();
    for (index, chunk) in entry.chunks.iter().enumerate() {
        payloads.push(collect_summary::request(
            ai,
            &crate::collect_prompts::chunk(entry, chunk, index, mode),
            timeout,
            mode,
            Context {
                label: format!("{uri} chunk {}/{}", index + 1, entry.chunks.len()),
                phase: "chunk_summary",
                uri: Some(&uri),
                chunk: Some(index + 1),
                chunks: Some(entry.chunks.len()),
            },
            logger,
        )?);
        let _ = sender.send(Event::Chunk);
    }
    let mut merged = collect_summary::merge(&payloads, mode);
    if payloads.len() > 1 && collect_summary::needs_compression(&merged) {
        match collect_summary::request(ai, &crate::collect_prompts::merge(&uri, &payloads, "session", mode), timeout, mode, Context {
            label: format!("{uri} session merge"), phase: "session_merge", uri: Some(&uri), chunk: None, chunks: Some(entry.chunks.len()),
        }, logger) {
            Ok(summary) => merged = summary,
            Err(error) => logger.log("llm_merge_fallback", json!({"phase":"session_merge", "session_uri":uri, "chunk_total":entry.chunks.len(), "error":error.to_string()})),
        }
    }
    Ok(merged)
}

pub fn run(
    entries: &[Entry],
    ai: &AiConfig,
    config: &CollectConfig,
    mode: Mode,
    logger: &Logger,
    zh: bool,
    warnings: &mut impl Write,
) -> crate::Result<(Vec<Group>, usize, usize)> {
    let total_chunks: usize = entries.iter().map(|e| e.chunks.len()).sum();
    let chunk_values = |count: usize| {
        [
            ("current", count.to_string()),
            ("total", total_chunks.to_string()),
            ("concurrency", config.concurrency.to_string()),
        ]
    };
    let session_values = |count: usize| {
        [
            ("current", count.to_string()),
            ("total", entries.len().to_string()),
        ]
    };
    progress(
        "COLLECT_PROGRESS_SUMMARIZE_CHUNKS",
        &chunk_values(0),
        zh,
        warnings,
    )?;
    progress(
        "COLLECT_PROGRESS_MERGE_SESSIONS",
        &session_values(0),
        zh,
        warnings,
    )?;
    let (sender, receiver) = mpsc::channel();
    let next = AtomicUsize::new(0);
    let mut results = vec![None; entries.len()];
    let mut failed = 0;
    let mut last_error = None;
    std::thread::scope(|scope| -> crate::Result<()> {
        for _ in 0..config.concurrency.min(entries.len()) {
            let sender = sender.clone();
            let next = &next;
            scope.spawn(move || {
                loop {
                    let index = next.fetch_add(1, Ordering::Relaxed);
                    let Some(entry) = entries.get(index) else {
                        break;
                    };
                    let result = summarize(entry, ai, config.timeout, mode, logger, &sender);
                    if sender.send(Event::Complete(index, result)).is_err() {
                        break;
                    }
                }
            });
        }
        drop(sender);
        let mut chunks = 0;
        let mut merged = 0;
        for event in receiver {
            match event {
                Event::Chunk => {
                    chunks += 1;
                    progress(
                        "COLLECT_PROGRESS_SUMMARIZE_CHUNKS",
                        &chunk_values(chunks),
                        zh,
                        warnings,
                    )?;
                }
                Event::Complete(index, Ok(summary)) => {
                    results[index] = Some(summary);
                    merged += 1;
                    progress(
                        "COLLECT_PROGRESS_MERGE_SESSIONS",
                        &session_values(merged),
                        zh,
                        warnings,
                    )?;
                }
                Event::Complete(index, Err(error)) => {
                    failed += 1;
                    let uri = entries[index].uri();
                    writeln!(
                        warnings,
                        "{}",
                        crate::i18n::terminal(
                            "WARN_SESSION_SUMMARY_SKIPPED",
                            zh,
                            &[("uri", uri.clone()), ("error", error.to_string())]
                        )
                    )?;
                    logger.log(
                        "session_summary_failed",
                        json!({"session_uri":uri, "error":error.to_string()}),
                    );
                    last_error = Some(error);
                }
            }
        }
        Ok(())
    })?;
    let included = entries.len() - failed;
    if included == 0
        && let Some(error) = last_error
    {
        return Err(error);
    }
    if failed > 0 {
        writeln!(
            warnings,
            "{}",
            crate::i18n::t(
                "WARN_SESSION_SUMMARY_FAILURES",
                zh,
                &[("count", failed.to_string())]
            )
        )?;
    }
    progress(
        "COLLECT_PROGRESS_RENDER_FINAL",
        &[("current", "0".into()), ("total", "2".into())],
        zh,
        warnings,
    )?;
    let mut working: Vec<Group> = entries
        .iter()
        .zip(results)
        .filter_map(|(entry, summary)| {
            summary.map(|summary| Group {
                date: entry.date.to_string(),
                project_directory: entry.session.working_directory(),
                session_uris: vec![entry.uri()],
                summary,
            })
        })
        .collect();
    let mut depth = 0;
    loop {
        let mut buckets: Vec<Vec<Group>> = Vec::new();
        let mut positions = HashMap::new();
        for group in &working {
            let scope = if mode == Mode::Insight || group.project_directory.is_empty() {
                group.session_uris[0].as_str()
            } else {
                ""
            };
            let key = (&group.date, &group.project_directory, scope);
            let index = *positions.entry(key).or_insert_with(|| {
                buckets.push(Vec::new());
                buckets.len() - 1
            });
            buckets[index].push(group.clone());
        }
        if buckets.len() == working.len() {
            break;
        }
        depth += 1;
        let total: usize = buckets.iter().map(|b| b.len().div_ceil(8)).sum();
        let values = |current: usize| {
            [
                ("level", depth.to_string()),
                ("current", current.to_string()),
                ("total", total.to_string()),
            ]
        };
        progress("COLLECT_PROGRESS_TREE_REDUCTION", &values(0), zh, warnings)?;
        let mut next = Vec::new();
        for bucket in buckets {
            for batch in bucket.chunks(8) {
                let payloads: Vec<_> = batch.iter().map(|g| g.summary.clone()).collect();
                let mut merged = collect_summary::merge(&payloads, mode);
                let index = next.len() + 1;
                let source = format!("collect://group-level-{depth}/group-{index}");
                if batch.len() > 1 && collect_summary::needs_compression(&merged) {
                    match collect_summary::request(ai, &crate::collect_prompts::merge(&source, &payloads, &format!("group-level-{depth}"), mode), config.timeout, mode, Context { label:source.clone(), phase:"group_merge", uri:None, chunk:None, chunks:None }, logger) {
                        Ok(summary) => merged = summary,
                        Err(error) => logger.log("llm_merge_fallback", json!({"phase":"group_merge", "context":source,"level":depth,"group_index":index,"error":error.to_string()})),
                    }
                }
                next.push(Group {
                    date: batch[0].date.clone(),
                    project_directory: batch[0].project_directory.clone(),
                    session_uris: batch.iter().flat_map(|g| g.session_uris.clone()).collect(),
                    summary: merged,
                });
                progress(
                    "COLLECT_PROGRESS_TREE_REDUCTION",
                    &values(index),
                    zh,
                    warnings,
                )?;
            }
        }
        working = next;
    }
    progress(
        "COLLECT_PROGRESS_RENDER_FINAL",
        &[("current", "1".into()), ("total", "2".into())],
        zh,
        warnings,
    )?;
    Ok((working, depth, included))
}
